### ISSUE TITLE
Basic react functional component widget doesn't work

### ISSUE THREAD

**Issue by faissalbyld** [annotorious/annotorious#202]
Hello, I'm using Annotorious 2.7.1, I was trying to do a minimal example of a React widget but it doesn't seem to work and I get the error: `TypeError: can't define property "__": Object is not extensible`
I have tried defining the widget like this:
```
const TestWidget = (props) => <div>Test widget</div>
```
And like this:
```
class UnitWidget extends Component {
  render() {
    return <div>Test widget</div>
  }
}
```

and I'm initializing Anno with:
```
      annotorious = new Annotorious({
        image,
        widgets: [{widget: TestWidget, force: 'react'}]
      });
```

I have tried the plainjs example from https://recogito.github.io/guides/editor-widgets/ and that works fine, the problem seems only to be when I try to use the React component functionality. I'm sort of fumbling in the dark/source code since I couldn't find any docs on using this with react, so I might just be missing something and would appreciate any help

**Comment by rsimon** [comment]
> I couldn't find any docs on using this with react

Yes, there is no good documentation around this at all yet :-(

The good news is: the component is correct, and you can use class and functional components. It's the build process that's the tricky part. Can you say a bit about how you are building the project (Webpack?) and what your build/Webpack config looks like?

There is also a thread on the Gitter chat:
https://gitter.im/recogito/annotorious?at=61f73384d41a5853f9781b90 

It's about the exact same problem, and might have a working solution. (Click on the '14 replies' link to open the thread.)

**Comment by rsimon** [comment]
P.S.: Writing better documentation around this is *almost* next on my list. (I've started writing a tutorial series on Observable - see [here](https://observablehq.com/@rsimon/annotorious-intro) and [here](https://observablehq.com/@rsimon/writing-annotorious-formatters-pt-1).) I'm currently finishing another tutorial on advanced formatters, and then widget development will be next.

**Comment by faissalbyld** [comment]
I'm doing this from within a basic CreateReactApp v5.0.0 so no webpack config changes beyond that.

I tried using the `config-overrides` package with the following config based on that thread you mentioned as I understood it:
```
const path = require('path');
module.exports = {
    webpack: function(config, env) {
      config.resolve.alias.react = path.resolve('./node_modules/react')
      config.resolve.alias['react-dom'] = path.resolve('./node_modules/react-dom')
        return config;
    }
}
```

But it just gave me these errors:
```
Compiled with problems:


ERROR in ./src/App.js 6:0-63

Module not found: Error: You attempted to import <path>/node_modules/react which falls outside of the project src/ directory. Relative imports outside of src/ are not supported.
You can either move it inside src/, or add a symlink to it from project's node_modules/.


ERROR in ./src/App.js 11:0-58

Module not found: Error: You attempted to import <path>/node_modules/react/jsx-dev-runtime which falls outside of the project src/ directory. Relative imports outside of src/ are not supported.
You can either move it inside src/, or add a symlink to it from project's node_modules/.


ERROR in ./src/index.js 4:0-26

Module not found: Error: You attempted to import <path>/node_modules/react which falls outside of the project src/ directory. Relative imports outside of src/ are not supported.
You can either move it inside src/, or add a symlink to it from project's node_modules/.


ERROR in ./src/index.js 5:0-33

Module not found: Error: You attempted to import <path>/node_modules/react-dom which falls outside of the project src/ directory. Relative imports outside of src/ are not supported.
You can either move it inside src/, or add a symlink to it from project's node_modules/.


ERROR in ./src/index.js 9:0-58

Module not found: Error: You attempted to import <path>/node_modules/react/jsx-dev-runtime which falls outside of the project src/ directory. Relative imports outside of src/ are not supported.
You can either move it inside src/, or add a symlink to it from project's node_modules/.


```

**Comment by rsimon** [comment]
The essential bit is too have react and react-dom as a peer dependency. Therefore, you won't be able to build a plugin with CRA. (I guess you can eject, though and move them to peerDeps.) 

The thread should have a link to an example project.

**Comment by rsimon** [comment]
Wait... from the error messages or looks like the peer deps have nothing to do with it. (Also depends if you want to build the widget as a separate component, or create a React app which includes a widget.)

Either way: it's not currently possible to (easily) build widgets when using CRA.

**Comment by rsimon** [comment]
For an example React project that imports Annotorious and adds custom plugins, see here:
https://github.com/machines-reading-maps/mrm-recogito-ui

Note specifically the `package.json` config and `webpack.config.js`.

**Comment by rsimon** [comment]
For posterity:
- It is, unfortunately, __not__ possible to use React-based widgets in a CRA project. Reason: Annotorious/RecogitoJS requires a few extra build settings which are only possible if you eject your app
- When using Annotorious/RecogitoJS inside a React project, __you need to import Annotorious/RecogitoJS from source__. If you import the transpiled versions, you'll end up with two copies of React (or, to be precise) one copy of React from your project, and one copy or Preact from Annotorious. This will lead to a conflict and the error reported above.
- A reference project with a Webpack build is available here: https://github.com/rsimon/ga-probate-annotate

I'm summarizing some key bits about this reference project:

1. In [webpack.config.json](https://github.com/rsimon/ga-probate-annotate/blob/master/webpack.config.js#L23-L29) you'll need the following alias declaration 

```js
{
    alias: {
      'react': 'preact/compat',
      'react-dom': 'preact/compat'
    }
}
```
2. When importing Annotorious, RecogitoJS or any of the plugins, you __always__ need to import from source, [as shown here](https://github.com/rsimon/ga-probate-annotate/blob/master/src/App.tsx#L3-L4):

```js
import { Recogito } from "@recogito/recogito-js/src";
import GeotaggingWidget from "@recogito/geotagging-widget/src";
``` 

3. Because you are now importing from source, you also need to include the transient `devDependencies` needed by Annotorious/RecogitoJS into your [package.json](https://github.com/rsimon/ga-probate-annotate/blob/master/package.json) - and all the plugins you use:

- @babel/preset-env
- @babel/preset-react
- babel-loader
- css-loader
- sass
- sass-loader
- style-loader

**Comment by Shyiy** [comment]
Is there a way to actually use React widgets in a React project without completely changing how the project is being built and importing things from source? Switching via alias to always use Preact is not really an acceptable solution in most situations as it has consequences for all other files in the project, additionally importing the dependency from source would also require whatever bundler/transpile being used to support and transpile the dependencies using jsx/es6/whatnot, which for instance parceljs simply doesn't support (intentionally).

Using a simple self-contained component in a separate file seems to work, but as soon as a dependency from another local component or other dependency (like react-select) using React is used, things break down with the mentioned `TypeError: can't define property "__": Object is not extensible` or `DOMException: Failed to execute 'createElement' on 'Document': The tag name provided ('[object Object]') is not a valid name` errors.

Possibly a separate library build or options in the use of its components/options?

**Comment by rsimon** [comment]
Unfortunately, no. And I'll openly admit that the reason for this was my naive use & lack of knowledge of React back in the day I started Annotorious :-) 

It's still going to stay that clunky for a while. However: I've already started working on the next major release of Annotorious [here](https://github.com/annotorious/annotorious-v3). Probably still 6 months or so at least before this gets usable. But the good news is that framework integration will be __much__ easier with v3. For React & Svelte there will be framework wrappers out of the box. For other frameworks, I hope that others will chip in.

**Comment by TomDijkema** [comment]
Hi @rsimon ,

Really appreciate the work you are putting into this library, it works very well!

I unfortunately just ran into this issue within my own React project, built with CRA.
I was wondering if the new version of Annotorious, that includes better widget integration with frameworks, is also going to be supported for OpenSeaDragon?
The project I am working on is in a state where annotations have become an important part of the mvp and I am subjected to complete this soon.
I was wondering if you would advice to try and apply the work around mentioned in the discussion above or wait for the new release of Annotorious if this supports OpenSeaDragon?

Kind regards,

Tom

**Comment by rsimon** [comment]
Hi @TomDijkema, 

yes, the new version supports OpenSeadragon. Likewise: there is still only minimal documentation. But if you are willing to give it a try - here's pretty much the absolute minimal setup for using Annotorious 3, OpenSeadragon version, via React bindings:

https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/test/App.tsx

In principle, everything is contained in the `@annotorious/react` package, which you can install from the npm registry.

I'm currently using this package heavily myself in one of my own projects. The codebase for this is a little big as an "introductory example"... But if you want to dig in: it would definitely give you more hints from a real-world scenario.

https://github.com/recogito/recogito-client/blob/main/src/apps/annotation-image/ImageAnnotationDesktop.tsx

**Comment by TomDijkema** [comment]
Thanks! I am trying to recreate a simple version of your example, but run into the error: 'c is not a function', when trying to load the OpenSeaDragonAnnotator/viewer and AnnotoriousOpenSeadragonAnnotator.
I know its still in development, therefore I might not correctly understand it yet and be doing something wrong, but still it is interesting to see how this will change up the method that is used with Annotorious version 2.

**Comment by rsimon** [comment]
Did you wrap the application with the `<Annotorious>` tag? The error message sounds like the top-level context might be missing. (The link I posted was actually a bit misleading, because in addition to `App.tsx` you'll also need to check `index.ts` - which is where the top-level Annotorious context provider is added.

https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/test/index.tsx

**Comment by rsimon** [comment]
That being said: if anyone knows how I can generate proper sourcemaps for the React binding module, do let me know... This would make the error message far more useful. But so far, I haven't figured out the proper build config...

**Comment by TomDijkema** [comment]
Ah yes! You are correct, I was missing the top Annotorious component that wraps it all up.
Now the viewer works just like normal. Will continue trying to apply the annotation methods now.

Regarding sourcemaps, I am afraid I do not have any experience with them, seems like it has to do with the build configuration like you said, so that it generates more specific error logs.

**Comment by TomDijkema** [comment]
Another question if I may: how does Annotorious hook onto the OpenSeaDragion viewer in this latest version?
I see you define the `anno` variable from the `useAnnotator` hook and this does contain the Annotorious methods afterwards, but I can not seem to figure out how you bind it to OpenSeaDragon.
In version 2 this was done by: `Annotorious(viewer, config)`, can you explain how this changed?

**Comment by rsimon** [comment]
That's still a bit of a work in progress in terms of how the API works. Right now, you'd have to use the `<OpenSeadragonViewer>` component. See here: https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/test/App.tsx

That will handle the initialization by itself. Not sure it's a great approach TBH, because people might want to do various customization when they initialize the viewer. FWIW, you can get a ref to the Viewer that's created in the component, like so:

```
const ref = useRef<OpenSeadragon.Viewer>(null);
...
<OpenSeadragonViewer ref={viewer} options={OSD_OPTIONS} />
```

**Comment by rsimon** [comment]
That's still a bit of a work in progress in terms of how the API works. Right now, you'd have to use the `<OpenSeadragonViewer>` component, if possible. See here: https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/test/App.tsx

That will then handle the initialization by itself. What happens is that the `<OpenSeadragonViewer>` hands the viewer instance to the [OpenSeadragonAnnotator](https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/src/openseadragon/OpenSeadragonAnnotator.tsx), which then initializes Annotorious. The wording in the API has changed a bit, but it's still essentially the same process, and [happens here](https://github.com/annotorious/annotorious-v3/blob/main/packages/annotorious-react/src/openseadragon/OpenSeadragonAnnotator.tsx#L36). Once `anno` is initialized, it's stored as a React context, and is then available to any component that sits below the top-level `<Annotorious>` context provider component.

TBH I'm not sure it's a great approach yet, because people might want to do various customization on the viewer when they initialize it. (And hiding viewer initialization in <OpenSeadragonViewer> might be limiting.) But, FWIW, if you need access to the viewer object, you can at least get it via a ref, like so:

```
const ref = useRef<OpenSeadragon.Viewer>(null);
...
<OpenSeadragonViewer ref={viewer} options={OSD_OPTIONS} />
```

**Comment by TomDijkema** [comment]
I see, it makes more sense now! Got the annotations in the viewer, they were just out of bounds :).
Annotations are also draggable now I see, interesting, can you toggle this behavior?

Depends on what kind of customization you want to apply as user.
The current method might lack customization at first, but is relatively easy to get going with.
Curious about which options are incorporated by default, any way the ref object should do the trick.

Going to look at making annotations visible at all times (they are hidden at first glance), activating draw mode and triggering a custom 'widget' form for adding annotations next.

**Comment by rsimon** [comment]
The annotation shapes are editable by default. Was also the case for the previous version. Although it had a `readOnly` option which isn't yet implemented in Annotorious 3. What you can do is this:

Initialize Annotorious with the following option (should work just fine as a prop to the `<OpenSeadragonAnnotator>`):

```
pointerSelectAction={PointerSelectAction.HIGHLIGHT}
```

or, if you're not using TypeScript, simply

```
pointerSelectAction="HIGHLIGHT"
```

This has the effect of still triggering selection events when you click annotations, but it will not make them editable. (Other options are `PointerSelectation.NONE` which will make them completely transparent to events, or `PointerSelectAction.EDIT` which is the default and will make them editable on click-select.)

**Comment by rsimon** [comment]
Re custom widget: as you may have seen, there's no more default editor. Anntorious 3 is [headless](https://annotorious.github.io/guides/headless-mode/) by default. There's an [open issue to build a default editor that looks & feels compatible to the 2.7 popup window](https://github.com/annotorious/annotorious/issues/297), but that hasn't happened yet.

Instead - you may have already seen this - there's now the `<OpenSeadragonPopup>` helper component. You should just be able to wrap any custom content in there, and the component will handle the popup position for you. The popup gets a set of props handed over: basically the selected annotation, a boolean flag indicating whether the annotation should be editable according to the current rules context, and the OpenSeadragon viewer (for convenience).

For anything else - access to annotations etc, you can use the various state hooks provided via the [Annotorious context component](https://github.com/annotorious/annotorious/blob/3.x/packages/annotorious-react/src/Annotorious.tsx). Let me know if you get it working! (I think you'll be the first person apart from myself to try it ;-)

**Comment by TomDijkema** [comment]
That's a first for me :).

Ok, used the Tooltip to recreate the old popup window. The option to add your own custom elements in here is very useful and is just the solution we needed for this Github issue. The tooltip placement is a bit wacky at some times, liked that the version 2 one always was displayed at a fixed position below the annotation.

The pointerSelectAction method works for edit, but highlight does not show anything for me. Also by default all annotations are invisible, only becoming visible when I click on them with the edit method. My expectation would be that all annotations are visible using grey-ish borders like in version 2 and become yellow when highlighted. At first I thought this might have to do with a styling issue, but now I think it has to do with the annotations not being present in the DOM when not selected with edit mode. The <g> element does exist on the annotation layer, but it is not populated until you trigger the edit mode for an annotation. Is this on purpose?

**Comment by rsimon** [comment]
Great to hear you got it working! Consider the tooltip placement a work in progress ;-)

Concerning the annotations: hm, that's weird. They are definitely supposed to be visible by default. I think there might still be something wrong with the rendering config. Do you have a code sample you can share? And an example of the annotations you are loading?

**Comment by TomDijkema** [comment]
Sure, this is the current setup I was working on, basically the example you gave with some additions (I did not touch the CSS of the Annotorious classes):
```

/* Import Dependencies */
import { useEffect, useState } from 'react';
import KeycloakService from 'keycloak/Keycloak';
import { Formik, Form, Field } from 'formik';
import {
    AnnotoriousOpenSeadragonAnnotator,
    ImageAnnotation,
    OpenSeadragonAnnotator,
    OpenSeadragonPopup,
    OpenSeadragonViewer,
    PointerSelectAction,
    useAnnotator
} from '@annotorious/react';
import { Row, Col } from 'react-bootstrap';

/* Import Styles */
import styles from 'components/general/mediaTypes/mediaTypes.module.scss';


import Annotations from './annotations.json';


/* OpenSeaDragon Configuration */
const OSD_OPTIONS: OpenSeadragon.Options = {
    prefixUrl: 'https://cdn.jsdelivr.net/npm/openseadragon@3.1/build/openseadragon/images/',
    tileSources: {
        "type": 'image',
        "url": 'https://files.geocollections.info/medium/51/c7/51c72e17-a4b9-4a6c-ab95-c4cef02c476a.jpg'
    },
    crossOriginPolicy: 'Anonymous',
    gestureSettingsMouse: {
        clickToZoom: false
    }
};


const ImageViewer = () => {
    /* Base variables */
    const anno = useAnnotator<AnnotoriousOpenSeadragonAnnotator>();
    const [selectedAnnotation, setSelectedAnnotation] = useState<ImageAnnotation>();

    useEffect(() => {
        if (anno) {
            const annotations = [...Annotations];

            anno.setAnnotations(annotations);
        }
    }, [anno]);

    /* Method for when selecting an Annotation */
    const SelectAction = (annotation: ImageAnnotation) => {
        setSelectedAnnotation(annotation);

        return KeycloakService.IsLoggedIn() ? PointerSelectAction.EDIT : PointerSelectAction.HIGHLIGHT;
    }

    return (
        <div className="h-100">
            <OpenSeadragonAnnotator className="h-100"
                keepEnabled={true}
                pointerSelectAction={SelectAction}
            >
                <OpenSeadragonViewer
                    className="openseadragon h-100"
                    options={OSD_OPTIONS}
                />

                {selectedAnnotation &&
                    <OpenSeadragonPopup
                        popup={() => (
                            <div className={styles.annotoriousPopUp}>
                                <Row>
                                    <Col className="bgc-white px-3 py-2">
                                        {/* Annotation values */}
                                        <Row className={`${styles.annotoriousPopUpSection} pb-2`}>
                                            <Col>
                                                {KeycloakService.IsLoggedIn() ?
                                                    <Formik
                                                        initialValues={{
                                                            annotationValue: selectedAnnotation ? selectedAnnotation.bodies[0].value : '',
                                                        }}
                                                        enableReinitialize={true}
                                                        onSubmit={async (form) => {
                                                            await new Promise((resolve) => setTimeout(resolve, 500));

                                                            /* Submit new Annotation */
                                                            console.log(form);
                                                        }}
                                                    >
                                                        <Form>
                                                            <Field name="annotationValue"
                                                                type="text"
                                                                className="fs-4"
                                                            />
                                                        </Form>
                                                    </Formik>
                                                    : <>
                                                        <Row>
                                                            <Col>
                                                                <p className="fs-4"> {selectedAnnotation.bodies[0].value} </p>
                                                            </Col>
                                                        </Row>
                                                    </>
                                                }
                                            </Col>
                                        </Row>
                                        {/* Close button */}
                                        <Row className="mt-3">
                                            <Col className="d-flex justify-content-end">
                                                <button type="button"
                                                    className="primaryButton fs-4 px-3"
                                                    onClick={() => setSelectedAnnotation(null)}
                                                >
                                                    Close
                                                </button>
                                            </Col>
                                        </Row>
                                    </Col>
                                </Row>
                            </div>
                        )}
                    />
                }
            </OpenSeadragonAnnotator>
        </div>
    );
}

export default ImageViewer;
```

**Comment by TomDijkema** [comment]
I used your example annotations, for example:
```
[{
        "@context": "http://www.w3.org/ns/anno.jsonld",
        "id": "#a88b22d0-6106-4872-9435-c78b5e89fede",
        "type": "Annotation",
        "body": [
            {
                "type": "TextualBody",
                "value": "A comment..."
            }
        ],
        "target": {
            "selector": {
                "type": "FragmentSelector",
                "conformsTo": "http://www.w3.org/TR/media-frags/",
                "value": "xywh=pixel:0,0,200,200"
            }
        }
    }]
```

**Comment by rsimon** [comment]
Ok, weird. This all looks perfectly correct. Anything in the browser console when you call `anno.setAnnotations()`?

A small remark (although that doesn't relate to the bug): you don't need the `{selectedAnnotation &&` check. The Popup is hidden automatically if there is no selection.

**Comment by TomDijkema** [comment]
Right? Mostly information:
![image](https://github.com/annotorious/annotorious/assets/102812384/5f83d43b-999d-4417-80d1-6126f77fa915)

Tried reinstalling the package, but that unfortunately did not do the trick

**Comment by rsimon** [comment]
Looks all good, too. Ok, I'll dig into it later. Let's see if there's anything I can reproduce.

**Comment by rsimon** [comment]
Ah wait - looke to quickly... that "WebGL context was lost" message doesn't look familiar. So you're saying the annotations don't render initially, but then later start appearing, after you have interacted? Could be some kind of timing/rendering issue involved.

**Comment by TomDijkema** [comment]
Well they are there, only the outline is not visible. I only see the yellow highlight outline when I click on the invisible annotation and trigger `PointerSelectAction.EDIT`. If I trigger it with `PointerSelectAction.HIGHLIGHT`, the outline stays invisible. Correct me if I am wrong, but normally all annotations are visible by the grey outlines, becoming yellow when clicked on.

![image](https://github.com/annotorious/annotorious/assets/102812384/918c7ead-4370-4d47-a1f8-3697b5dc4a9f)

**Comment by rsimon** [comment]
Ok, I couldn't figure it out yet. Can you confirm you're importing the stylesheet from `@annotorious/react/annotorious-react.css` anywhere in your app?

**Comment by TomDijkema** [comment]
I am so sorry! I still had the old style sheet of version 2 imported instead of the new one.
Now the annotations are visible by default!

**Comment by rsimon** [comment]
Excellent! Good to hear it's working now.

If you have the time, do keep me posted about your progress. I'm really interest in hearing how it goes (and how people are making use of Annotorious!)

**Comment by TomDijkema** [comment]
I am now working on creating new annotations and then letting the user add a value to the annotation through the popup. I run into an issue with focus. When I create an annotation by dragging and letting go of the left mouse button, it triggers the popup, but somehow it does not register that the annotation has been selected, as where ever I click the popup immediately closes. I tried focusing my input element after the popup trigger, that works for typing but does not solve the need to press the save button, which is still not clickable as it closes the popup.

I was thinking of programmatically selecting the newly created annotation to have it focus, because what also works is to click on the new annotation once more after creating. In the old documentation I read about the `selectAnnotation` method, but I don't think this is already in the new version.

Do you maybe have any pointers on this?

**Comment by rsimon** [comment]
Can you describe the exact sequence of events? Clicking outside the popup would close it. (I'm thinking about adding an optional "modal" mode which could render the image underneath intert. But for now, outside click works like in the old Annotorious version. Click outside = cancel.) 

But click on the popup itself shouldn't normally close it. I.e. something must go wrong at some point. (The popup doesn't have a `pointer-events: none` CSS rule set or anything?)

You could compare with a popup I built for my current project. The screenshot shows the situation right after creating a new annotation: the annotation is in selected mode, the popup is open, focus is on the text field.

<img width="521" alt="Bildschirmfoto 2023-10-03 um 12 50 23" src="https://github.com/annotorious/annotorious/assets/470971/f5f4aefe-0385-4ca7-b4ea-141b9f678c82">

Focus handling can be tricky though, and you'll currently have to take care of that yourself (for now!) Perhaps there's something you can grab from my popup. The full Popup component [is here](https://github.com/recogito/recogito-client/blob/main/src/components/Annotation/Popup/Popup.tsx). But what's probably most useful is this piece of code, which does the focus management:

https://github.com/recogito/recogito-client/blob/main/src/components/Annotation/ReplyForm/ReplyForm.tsx#L44-L47

**Comment by rsimon** [comment]
Ah: another question do you have the `keepEnabled` mode for the drawing tool on or off?

**Comment by TomDijkema** [comment]
I had the keepEnabled mode on, but when I turned it off the results remains the same.
Apparently this may a browser dependent issue, I use Fire Fox, but when I tested it in Chrome it worked as expected.

Somehow in Fire Fox when you draw an annotation, drag an area for it, and let loose of the left mouse button to complete it; the mouse cursor remains in a selecting state. With selecting state I mean you can hover over text and images and will continue selecting them all as if you were holding down the left mouse button and selecting a bunch of text to copy for example. It does not matter where you click after it gets in this state as it will not recognize the element thus close it.

Thanks for the example, I will continue and try to get it working somehow.

**Comment by rsimon** [comment]
Ok, that's wild. I wasn't aware - but I can reproduce this on Firefox with my own popup, too. Regardless of where you click, it has the same effect as if clicking outside. It doesn't seem to happen __every__ time. But most of the time.

I'll definitely make a note about that on our own issue tracker, and will try to get to the bottom of it. Let me know if you can find any more hints that might point to the root cause.

**Comment by rsimon** [comment]
BTW: not sure if that helps. But the issue doesn't affect polygon drawing. (I think it's simply because when you draw a polygon, your last action is a click or double-click, not a drag.)

**Comment by TomDijkema** [comment]
Oh yes that makes sense! The click would prevent the selection from going on.

Another subject :), did you perhaps rename the `cancelSelected` listener? When the user did not enter a value in the input pop-up but closes it, I want to detect the closing and remove the empty annotation from the annotations layer.

**Comment by TomDijkema** [comment]
Nevermind, found the `selectionChanged` method, which should also work!

**Comment by rsimon** [comment]
Hi, I think I found the culprit about the Firefox issue. I'm not claiming that I understand it entirely. But, clearly, native selection behavior intervenes on Firefox (and doesn't, on Chrome). The solution is to disable selection on the Annotorious container element. I'm including this in the next relase. But, for now, you can apply the fix in your own stylesheet:

```css
.a9s-annotationlayer {
  -webkit-user-select:none; 
     -moz-user-select:none;
      -ms-user-select:none;
       -o-user-select:none;
          user-select:none;
}
```

(Interestingly, I had that same setting in the Annotorious 2.7 stylesheet. Although I don't remember ever running into this particular bug.)

**Comment by TomDijkema** [comment]
Interesting, good to hear you found it! I will try and apply this right away!
A coincidence perhaps :).

**Comment by TomDijkema** [comment]
I think I got most of it working, adding, editing and removing annotations.
We are still working on the API a bit, but when its done I should also be able to share my solution with you.

Many thanks for the help until now!
I hope more people will find this new version of the package useful for implementation within JavaScript frameworks!

**Comment by TomDijkema** [comment]
Hi @rsimon,

Today we deployed the latest version to our development version that includes Annotorious version 3.
Currently just for testing purposes, but you can check it out here: https://dev.dissco.tech/dm/TEST/FCY-N2H-5AQ (creating and editing annotations is locked behind login).

It seems to work fine, occasionally we have some visual bugs with the blue bounding box representing the annotation.
For example is does not move around with the annotation boundaries itself when editing; or remaining a static part of the image when removing an annotation (appears to be stuck on the image itself).

If you are curious about the code, its mainly stored in these two files:
https://github.com/DiSSCo/disscover/blob/main/src/components/general/mediaTypes/ImageViewer.tsx
https://github.com/DiSSCo/disscover/blob/main/src/components/general/mediaTypes/ImagePopup.tsx

**Comment by TomDijkema** [comment]
Hi,

After updating the package to the latest version (3.0.0-pre-alpha-49) I am running into some issues.
If I am correct the structure for loading annotations changed, body is now bodies and coordinates of the bounding box are provided by a geometry object instead of the value string.
Existing annotations are loading fine with this format and creating new annotations also works; but somehow I can not interact with the annotations that are present on the canvas (highlight or edit).
Suspected this to be something that changed in the pointer select action, but I can not find any notable changes in your example project, seems like no select action is called at all.

Do you maybe have any pointers on the latest update?

**Comment by rsimon** [comment]
Dear Tom,

yikes - of course. There was a breaking change. (Hopefully the last one before the official release goes out.)

As you diagnosed directly, Annotorious uses a slightly different data model internally. (This is to avoid constant string parsing- and serialization-operations that would happen if I used the Web Annotation model throughout.)

After long consideration, I decided to make the internal model the "canonical" one. (Frankly, most people never cared about the W3C standard, and were just constantly upset that they had to put so much effort into simple things, e.g. getting point coordinates out of a polygon annotation - which meant getting a field from the JSON that contained a string, which was actually an SVG element, parse that string, get the points attribute and so on...) Obviously, Annotorious *still* fully supports the Web Annotation standard - only now you need a small extra configuration option to enable a built-in crosswalk.

The following should fix things for you, if you want to keep using the Web Annotation model. Instead of initializing like so:

```jsx
<OpenSeadragonAnnotator className="h-100"
  keepEnabled={true}
  pointerSelectAction={SelectAction}>
```

you need to add the W3C Adapter:

```jsx
import { W3CImageFormat } from '@annotorious/react';

<OpenSeadragonAnnotator className="h-100"
  adapter={W3CImageFormat}
  keepEnabled={true}
  pointerSelectAction={SelectAction}>
```

Concerning the select action: hm, nothing has changed there. Does the browser console say anything?

**Comment by rsimon** [comment]
BTW: I now took a look at your dev system. Thanks for sharing - looks great. I love the sidebar design! 

The popup position looks weird though. It's supposed to stay close to the annotation box. I'll try to investigate. Might be that Annotorious isn't taking the offset of the OSD container element into account...

Concerning your problems: hm, does the browser console say anything when annotations get stuck? This definitely shouldn't happen. (I have seen the interface become sluggish during editing, which would cause the blue overlay lag behind the editable outline when you moved it. Although that won't normally happen below <1.000 annotations per image.)

**Comment by TomDijkema** [comment]
Ah ok, I can understand you chose to opt for the internal model.
Seems simple for the vector square compared to the free polygon shape.
I think we want to keep using the W3C one (although it is not very hard to restructure it), tried to import the `W3CImageFormat`, but it does not seem to be exported from the @annotorious/react package.
Do I need to include this separately?

Have not found why it neglects interaction with annotations on the canvas. Clicking does not trigger any function, so there are no additional messages in the console next to the default information.
Its strange because I am allowed to create annotations, but then it completely negates all interaction. Nothing seems to be in the way of the cursor.
I'll keep testing it.

Regarding the visual issues, hmmm, might be that I am resetting annotations or rerendering Annotorious on the OpenSeaDragon viewer. I use the annotorious.setAnnotations() a lot for rerendering the annotations.

**Comment by rsimon** [comment]
Hi @TomDijkema, oh - could be that the latest package published on npm doesn't yet export it. You should be able to import it directly from the base package `@annotorious/annotorious`. However, I'll almost certainly publish another update to the npm registry by Monday, and that will definitely include the re-export of `W3CImageFormat` right from the `@annotorious/react` package.

> Have not found why it neglects interaction with annotations on the canvas. 

Does this happen right from the start? Or does it only break occasionally at some point? `.setAnnotations()` may be involved, good point. So are you saying annotations you remove with `.setAnnotations()` remain on the canvas as inaccessible zombie annotations?

**Comment by abiswas97** [comment]
@rsimon @TomDijkema I think I've faced a similar issue with the popup not really being where it should be, but being consistently offset by a certain amount. In my case, this was in fact the offset of the OSD container. 

I'm on v2, but using the code for the v3 popup to be a headless popup in v2. Adding the offset of the viewer to the annotation bounding rect that the popup position is based on solved it for me -
```
  // Get the viewer bounding rect
  const viewerBounds = viewer.element.getBoundingClientRect();

  // Annotation coordinates -> viewer element coordinates
  // Assuming that the min/max coordinates of your annotation are determined from the annotation's target selector
  const topLeft = viewer.viewport.imageToViewerElementCoordinates(
    new OpenSeadragon.Point(minX, minY)
  );
  const bottomRight = viewer.viewport.imageToViewerElementCoordinates(
    new OpenSeadragon.Point(maxX, maxY)
  );

  // The return here is the adjusted bounding rect of the annotation in the DOM. 
  return {
    x: topLeft.x + viewerBounds.x,
    right: bottomRight.x + viewerBounds.x,
    y: topLeft.y + viewerBounds.y,
    bottom: bottomRight.y + viewerBounds.y,
    width: bottomRight.x - topLeft.x,
    height: bottomRight.y - topLeft.y,
  };
  ```

**Comment by TomDijkema** [comment]
> Hi @TomDijkema, oh - could be that the latest package published on npm doesn't yet export it. You should be able to import it directly from the base package `@annotorious/annotorious`. However, I'll almost certainly publish another update to the npm registry by Monday, and that will definitely include the re-export of `W3CImageFormat` right from the `@annotorious/react` package.
> 
> > Have not found why it neglects interaction with annotations on the canvas.
> 
> Does this happen right from the start? Or does it only break occasionally at some point? `.setAnnotations()` may be involved, good point. So are you saying annotations you remove with `.setAnnotations()` remain on the canvas as inaccessible zombie annotations?

Nice! Then I'll await the new version that includes the export.

It breaks after I use the setAnnotations method, as the annotations seem to work when just using create and then reselecting them. Might be on a trace of something, will continue tomorrow and let you know what I can find!

**Comment by TomDijkema** [comment]
> @rsimon @TomDijkema I think I've faced a similar issue with the popup not really being where it should be, but being consistently offset by a certain amount. In my case, this was in fact the offset of the OSD container.
> 
> I'm on v2, but using the code for the v3 popup to be a headless popup in v2. Adding the offset of the viewer to the annotation bounding rect that the popup position is based on solved it for me -
> 
> ```
>   // Get the viewer bounding rect
>   const viewerBounds = viewer.element.getBoundingClientRect();
> 
>   // Annotation coordinates -> viewer element coordinates
>   // Assuming that the min/max coordinates of your annotation are determined from the annotation's target selector
>   const topLeft = viewer.viewport.imageToViewerElementCoordinates(
>     new OpenSeadragon.Point(minX, minY)
>   );
>   const bottomRight = viewer.viewport.imageToViewerElementCoordinates(
>     new OpenSeadragon.Point(maxX, maxY)
>   );
> 
>   // The return here is the adjusted bounding rect of the annotation in the DOM. 
>   return {
>     x: topLeft.x + viewerBounds.x,
>     right: bottomRight.x + viewerBounds.x,
>     y: topLeft.y + viewerBounds.y,
>     bottom: bottomRight.y + viewerBounds.y,
>     width: bottomRight.x - topLeft.x,
>     height: bottomRight.y - topLeft.y,
>   };
> ```

Interesting! I will try this out!

**Comment by rsimon** [comment]
@abiswas97 excellent, thanks. I guess that code could also go into the [OpenSeadragonPopup component](https://github.com/annotorious/annotorious/blob/3.x/packages/annotorious-react/src/openseadragon/OpenSeadragonPopup.tsx). I'll see where that might fit!

@TomDijkema I think I have a hunch what might be happening. In fact I faced a similar problem with the [Recogito Text Annotator](https://github.com/recogito/text-annotator-js), Annotorious' text annotation sister project. The problem was annotations not getting properly removed from the 2D index that keeps the annotation state. I hadn't encountered the problem in Annotorious so far. But then I believe I never actually used .setAnnotations except on an empty state.

Anyways: I just added the fix from Text Annotator to Annotorious. Fingers crossed that this might do the trick! Will also be part of the next package release.

**Comment by abiswas97** [comment]
@rsimon yup I'm using the OpenSeadragonPopup component, and this was a small modification made to the `getAnnotationDomRect` function in the `setPosition` file. Should I open a PR for this?

**Comment by rsimon** [comment]
Yes, that would be perfect thanks!

**Comment by rsimon** [comment]
I just published a new Annotorious 3 (pre-)release with all the latest fixes, including @abiswas97's popup fix (thanks again!) and a fix that may (or may not...) solve your problem with stray annotations, @TomDijkema.

https://www.npmjs.com/package/@annotorious/react/v/3.0.0-pre-alpha-50

**Comment by TomDijkema** [comment]
The pop-up is much better now! It hovers just under the annotation.

I added the W3CImageFormat, it initializes, but gives me an error when I try to call the `getAnnotations()` function to check if there are any annotations present: `e.serialize is not a function`.
When I do not use the adapter it works fine and returns the expected empty array.

I am hoping switching back to W3C will resolve the annotations which can not be interacted with, because I think I might be doing something wrong with new geometry object and it does not register the correct bounds or something, hence I can not click on the area to trigger the highlight or edit.

**Comment by rsimon** [comment]
@TomDijkema sorry, should have checked my own docs before writing the instructions ;-)

My example was wrong! Here's how it works:

```jsx
<OpenSeadragonAnnotator 
  adapter={W3CImageFormat('https://iiif.bodleian.ox.ac.uk/iiif/image/af315e66-6a85-445b-9e26-012f729fc49c')}>
          
  <OpenSeadragonViewer className="openseadragon" options={OSD_OPTIONS} />

</OpenSeadragonAnnotator>
```

In short, `W3CImageFormat` is a function, and you supply it with a single argument. This argument is used as the `source` attribute for the W3C Web Annotation target. (The internal data model doesn't have a concept of a `source`, and it's definition will depend on the type of image, anyway. Therefore, I thought it's actually useful to have it configurable from the outside.)

You can, technically, omit the argument, and just use `W3CImageFormat()`. In this case, the `source` field will still be included in the Web Annotation, but with an `undefined` value.

**Comment by TomDijkema** [comment]
Yes it works now!
I think this is a great solution to keep it separate but still offering the option to the user if they want to.

Using the W3C standard makes selecting the annotations work fine again :+1: 
The issue only remains when removing annotations.
It seems Annotorious still expects the annotation to be there, while it isn't.
I remove annotations by removing it from the array and resetting the state with `setAnnotations`
Tried to also use `clearAnnotations` or `removeAnnotation`, but the issue persists.

Several seconds after deleting or when trying to use any other method of Annotorious it gets stuck and givers either:
`Uncaught TypeError: G is undefined` or
`I is undefined`

Update: it only seems to happen if I delete an annotation that was not created during the same Annotorious session.
So if I create an annotation, click on it and remove it, it works as expected, removing the annotation and the blue box.
If if create an annotation, then first refresh the page and then try to remove it, it seems to be removed from the state, but the blue bounding box will stay and it gives the error mentioned above when you want to interact with it.

**Comment by TomDijkema** [comment]
On load of the page, I set existing annotations with the `setAnnotations` function. I expect this to be the same as if you would manually add annotations, but maybe the problem has to do with that somehow.

**Comment by rsimon** [comment]
Argh. Sorry - I can see where the problem is. Individual deletes (anno.removeAnnotation) should work though?

Anyways: expect a fix coming your way.

**Comment by TomDijkema** [comment]
Hmm if I use the anno.removeAnnotation method it does delete the annotation from the internal state (tested with the getAnnotations method), but it recreates the same issue with the blue bounding box remaining, giving the error when trying to interact with it.

Really appreciate the effort!

**Comment by rsimon** [comment]
Ok, I just published another version: 3.0.0-pre-alpha-51 - make sure to grab that one.

That definitely fixed the problem with `anno.clearAnnotations()` and `anno.setAnnotations()` on my end. I couldn't reproduce the problem with `.removeAnnotation(id: string)` though. Could it be that you had called `.removeAnnotation()` on an annotation that was no longer in the state (because of running clear-/setAnnotations previously), but still on the screen because of the bug?

Either way: let me know if the new version improves anything.

**Comment by TomDijkema** [comment]
Hi, thanks! The update fixed the issue with deleting annotations when the page is refreshed.
Somehow it did create another issue. When I create an annotation, edit it (within the same session) and try to delete it, the dangling bounding box issue persists.
I also found the dangling bounding box to be present when adding or editing an annotation on an image, then switching to another image with a completely separate Annotorious initialization (change image calls on new Annotorious).

What's so strange about this dangling bounding box issue is that if I check the current annotations of the Annotorious session, it's not there, except visually it is.
This is especially strange when a completely different Annotorious object is created :thinking:

**Comment by rsimon** [comment]
Phew... this is really weird. The store isn't a global thing, i.e. it should be perfectly fine to have multiple instances in parallel. However, it doesn't yet have a proper `.destroy()` method. Are you creating a new Annotorious instance over an existing one? (And could it be that you then end up having two WebGL `<canvas>` elements on top of OpenSeadragon? (With the first one frozen to the last state of the old Annotorious instance?)

With regard to deletion: I can't reproduce that in the basic setup (just `@annotorious/react` with OpenSeadragon). Do you think it might be possible for you to set up a minimal online example where this is reproducible (ideally a Stackblitz/Codesandbox kind of thing?)

**Comment by TomDijkema** [comment]
Hmmm that might be the case, I will have to check that out.

I suspect the deletion might have to do something with how I treat the annotations.
Almost got an online version running, but some last things don't work well, I already have it too embedded.
For now I have to continue with other work first, but when I have the time to pick this up, I will let you know the outcomes!

**Comment by rsimon** [comment]
Ok, thanks. Come to think of it... I could imagine there are several situations where the React `<OpenSeadragonAnnotator>` gets remounted. Because there's no .destroy method yet that could get called when unmounting, this might lead to stray annotations. I'll investigate in any case, and add a destroy() method + proper unmounting for the next package release.

**Comment by TomDijkema** [comment]
Hi @rsimon,

It has been some time and I do not know if this issue is still the best place to ask, but I was wondering, do you have any trouble loading images from webservers into the OpenSeaDragon canvas?
I finally got some time to look at the Annotorious part again and noticed that all images are getting blocked due to a CORS issue.
The people from OpenSeaDragon suggested trying to disable the pyramid build option, but I am unable to identify this option within the typing of the OpenSeaDragon options or tile sources objects as defined by Annotorious.

Just wondering if something comes to your mind about this issue?
The issue from OpenSeaDragon I opened: https://github.com/openseadragon/openseadragon/issues/2444

**Comment by rsimon** [comment]
Are you getting this issue if you use OpenSeadragon __only__? Or is this something that specifically happens with Annotorious? I never had problems with OSD directly, and I'd be surprised - because IMG requests wouldn't normally fall under the CORS security policy. 

The current alpha builds of Annotorious, on the other hand, have a little quirk which will complain about cross-domain issues. Either way, it may be worth a try to set `crossOriginPolicy: "Anonymous"` in your OpenSeadragon config:

```
const OSD_OPTIONS = {
  prefixUrl:
    "https://cdn.jsdelivr.net/npm/openseadragon@4.1/build/openseadragon/images/",
  tileSources: {
    type: "image",
    url: "/1280px-Hallstatt.jpg",
  },
  crossOriginPolicy: "Anonymous",
  gestureSettingsMouse: {
    clickToZoom: false,
  },
};
```

If that doesn't resolve your problem, then I fear there is no way around getting the image server to set CORS headers. (Or building your own server-side proxying.)

**Comment by rsimon** [comment]
Yikes, I see in the thread over at OSD that this is already what you did... Can you paste the exact error message, so I can see if it's related to the quirk I mentioned earlier. (Incidentially, the fix for this is something [I did just this morning](https://github.com/annotorious/annotorious/issues/328). But it will still ~1 week or two until I can publish next release candidate that includes the fix.

**Comment by rsimon** [comment]
BTW: another thing. If you can load the image into an `<img>` tag, there might be a way to parse it into a data URL somehow? If that's possible, you can hand the image data URL to OpenSeadragon and it will display it normally. I'm currently doing something similar in another project. (In this case, I'm reading the image from a local file to a data URL, [and then hand that to OSD](https://github.com/rsimon/immarkus/blob/main/src/pages/annotate/WorkspaceSection/AnnotatableImage.tsx#L38).)

**Comment by TomDijkema** [comment]
I am really interested in the method you are describing about using a data URL.
Can I ask how the image object in the project is structured and what for example the values are?
If I could solve it that way, that would be great.

For your question about the exact errors I receive:
- If I use the CrossOriginPolicy, it gives: `Unable to open [object Object]: Error loading image at https://mediaphoto.mnhn.fr/media/1681821627482HLYgnWjzyeRM9kvB`
- When not using CORS, it loads the image but gives: `Uncaught DOMException: The operation is insecure.` and renders the canvas unmutable (resizing, zooming, etc.) but the Annotorious annotations seem to keep working

In the OSD thread they suggested setting the `buildPyramid` option to fals,e, but TypeScript fails to find this option on the types, otherwise disabling typing and applying the option, it sadly does not change anything

**Comment by rsimon** [comment]
In this particular app, the `Image` is just [a custom type that I defined](https://github.com/rsimon/immarkus/blob/main/src/model/Image.ts) The important part is the `data` field. That contains the binary data, [read from a file on the user's machine](https://github.com/rsimon/immarkus/blob/main/src/store/utils.ts#L1-L16).

Once you have the binary data, you can just stuff it into a data URL [like shown above](https://github.com/rsimon/immarkus/blob/main/src/pages/annotate/WorkspaceSection/AnnotatableImage.tsx#L38). In your case, I assume you'd rather want to pull the binary data from an HTML image element in the DOM. I.e. you'd probably looking for something like this:
https://stackoverflow.com/questions/42471755/convert-image-into-blob-using-javascript

The StackOverflow suggests two options: either via fetch (which would mean it falls under the CORS restrictions), or by copying the image into a Canvas element. I'm not 100% sure this will work though... the browser might still apply the cross-domain restrictions as soon as you attempt the copy.

FWIW: your error message is puzzling... I wonder if it's really a CORS issue. Did you check if the HTTP image response is correct? Similar errors could happen if the server returns an incorrect MIME type, for example.

**Comment by rsimon** [comment]
Ok, wait: I think this is what's going on. I tried to open your example image above. Turns out it's not available as HTTPS - only HTTP! So instead of CORS, you might be running into a security issue because you are fetching an unsecure resource (the image) into a secure JS app (you app).

This would match the error characteristics above I'd say. Not that this insight matters much - because the browser will enforce the exact same limits in this case as for CORS. I believe this also rules out the two StackOverflow solutions above. Unless you can get the image server to offer HTTPS access, building a simple server-side proxy is probably your only option. (I remember doing something similar once... can't remember where/when exactly. But I believe I've used some kind of [Netlify redirect](https://docs.netlify.com/routing/redirects/rewrites-proxies/) which was fairly simple to set up & free.)

**Comment by TomDijkema** [comment]
Yeah the image server itself is probably the issue and the solution, will have to contact the people behind it.
Either way, I still tried the converting option, but indeed that option fails due to the fetch getting the same CORS based error.
The proxy suggestion could work, we do have an unforeseeable amount of image servers to include, so we will have to make a strategy for this.

Thanks again for thinking with me!

**Comment by TomDijkema** [comment]
Hi!

I managed to spent time on the Annotorious implementation and pretty much revisited all the code.
Using the latest release candidate (17), I was able to refactor the code to resolve most of the issues with stale annotations and errors we spoke about earlier.
Sometimes it generated an error like: 'F is not defined', and I think this is because it was trying to delete or patch an annotation, but not being able to find it because Annotorious had lost trace of it.
This was probably due to me creating a lot of external state variables to keep track of the annotations, which I now see is much better to let Annotorious handle that itself and pull data from the available API calls.

Really satisfied with how its shaping up! Are you planning a final release candidate sometime soon?

I encountered one more issue with the typing of the W3CAnnottion type. Type script does not recognize any properties upon the target within the W3CAnnotation type, aka the W3CAnnotationTarget.
It is there in GitHub: https://github.com/annotorious/annotorious/blob/main/packages/annotorious-core/src/model/W3CAnnotation.ts, so maybe something goes wrong when exporting the type interface?
When I use ts-nocheck, it does work.

I also wondered if I can make the pointerSelectAction attribute upon the OpenSeaDragonAnnotator return a W3CAnnotation instead of an ImageAnnotation?
To note: I initialized annotorious with the W3CImageFormat adapter and used the W3CAnnotation in the annotorious variable like so: `useAnnotator<AnnotoriousImageAnnotator<W3CAnnotation>>()`

Best regards,

Tom

**Comment by rsimon** [comment]
Hi,

that's all great to hear. The release candidates are getting pretty stable now. Incremental enhancements rather than breaking changes! Things are basically ready to go, I'm just hesitant to put a real "v3.0.0" label on them before there is...
- proper documentation...
- an add-on package that emulates the built-in popup from v2.7, and provides a simple migration path for those who are using v2 with the popup now

This means I still can't say exactly when launch day will be. But at least, releases should become pretty stable from now on.

Concerning the TypeScript issues.
- Hm, the annotation types are in `@annotorious/core`. Can you check your `node_modules` folder if the `core` module is actually there, in the `@annotorious` folder? If not, you could try deleting `node_modules` and `package-lock`, and running a fresh `npm install`.
- Good point about the `pointerSelectAction`. Yes, I think it would make sense to run that through the adapter. I created an issue for this here: #357

**Comment by TomDijkema** [comment]
That sounds about right.
If I can find some time I could assist you with some documentation about the React implementation etc. as I really appreciate the effort going into this, but at the moment quite busy :|.

I checked the node modules folder and its in there, it also indicates the type of the target is W3CAnnotationTarget, but somehow has no properties (which should be source and selector right).
Removing and reinstalling the node modules sadly did not fix it.

**Comment by rsimon** [comment]
Ah, wait. The value of `target` can be a `W3CAnnotationTarget` or an array `W3CAnnotationTarget[]` (as per the W3C spec). If you know that it's a single target, does `as W3CAnnotationTarget` help?

**Comment by TomDijkema** [comment]
Ah yes, I changed it to make sure its a single W3CAnnotationTarget, it recognizes the selector within it, but complains about the lack of the value field within the selector to get the coordinates from.
I think the AbstractW3CSelector should take care of this? The props including the value property are present in a return statement from the parseW3CBodies function, but not sure how it would append them to the type definition?

Any way, for now I resolved it (quickly) like this:
```
const target = annotoriousAnnotation.target as W3CAnnotationTarget & { selector: { value: string } };
const coordinates = target.selector?.value.split(',');
```

**Comment by rsimon** [comment]
I think you might have to type your annotator in the following way:

```
useAnnotator<AnnotoriousImageAnnotator<W3CImageAnnotation>>()
```

(Note that it's now `W3CImageAnnotation`, rather than `W3CAnnotation`.) The W3C spec is pretty nested & flexible. Not all selectors have values. The image selectors do, though. Once you type your annotation as a W3CImageAnnotation, the value field should be there. (Fingers crossed.)

**Comment by TomDijkema** [comment]
That works! Had to download the regular annotorious/annotorious 3.0.0-rc.18 package for it because the types of W3CImageAnnotation are not in the React package.

Also had to set the target to the W3CImageAnnotationTarget type, and the selector to the W3CImageAnnotationSelector type.

Result:
```
const target = annotoriousAnnotation.target as W3CImageAnnotationTarget;
const selector = target.selector as W3CImageSelector;
const coordinates = selector?.value.split(',');
```

**Comment by rsimon** [comment]
Great! 

I added W3CImageAnnotation and W3CImageAnnotationTarget as re-exports to the `@annotorious/openseadragon` and `@annotorious/react` packages. So installing `@annotorious/annotorious` won't be necessary anymore from the next release onwards.

I think there won't be a way around using `as W3CImageAnnotationTarget` though. `W3CImageAnnotation`s are allowed to have either a single target, or a list, which is why TS can't infer. (I guess you could do something like `if (Array.isArray(target)) { ... }` though.)
